filebeat/input/filestream/input.go

```diff
@@ -168,7 +184,160 @@ func (inp *filestream) Run(
 
 	// The caller of Run already reports the error and filters out errors that
 	// must not be reported, like 'context cancelled'.
-	return inp.readFromSource(ctx, log, r, fs.newPath, state, publisher, metrics)
+	err = inp.readFromSource(ctx, log, r, fs.newPath, state, publisher, metrics)
+	if err != nil {
+		// First handle actual errors
+		if !errors.Is(err, io.EOF) && !errors.Is(err, ErrInactive) {
+			return fmt.Errorf("error reading from source: %w", err)
+		}
+
+		if inp.deleterConfig.Enabled {
+			if err := inp.deleteFile(ctx, log, cursor, fs); err != nil {
+				return fmt.Errorf("cannot remove file '%s': %w", fs.newPath, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// waitGracePeriod waits for the delete grace period while monitoring the file
+// for any changes, if the file changes or any error is encountered, false
+// is returned. True is only returned if the grace period expires with
+// no error and no context cancellation.
+func (inp *filestream) waitGracePeriod(
+	ctx input.Context,
+	logger *logp.Logger,
+	cursor loginp.Cursor,
+	fs fileSource,
+) (bool, error) {
+	// Check if file grows during the grace period
+	// We know all events have been published because cursor.Finished is true,
+	// so we can get the offset from the cursor and compare it with the file size.
+	st := state{}
+	if err := cursor.Unpack(&st); err != nil {
+		return false, fmt.Errorf("cannot unpack cursor from '%s' to read offset: %w",
+			fs.newPath,
+			err)
+	}
+
+	graceTimer := time.NewTimer(inp.deleterConfig.GracePeriod)
+	checkIntervalTicker := time.NewTicker(inp.scannerCheckInterval)
+	// Wait for the grace period or for the context to be cancelled
+LOOP:
+	for {
+		select {
+		case <-ctx.Cancelation.Done():
+			return false, ctx.Cancelation.Err()
+		case <-checkIntervalTicker.C:
+			stat, err := os.Stat(fs.newPath)
+			if err != nil {
+				// Return the error and cause the harvester to close
+				return false, fmt.Errorf("cannot stat '%s': %w", fs.newPath, err)
+			}
+
+			if stat.Size() != st.Offset {
+				// The file has changed, close the harvester
+				// without deleting the file
+				logger.Debugf("cancelling deletion of '%s', size has changed from %d to %d",
+					fs.newPath,
+					st.Offset,
+					stat.Size())
+				return false, nil
+			}
+		case <-graceTimer.C:
+			break LOOP
+		}
+	}
+
+	stat, err := os.Stat(fs.newPath)
+	if err != nil {
```

Comment:
Same here, something else removing the file isn't an error if it's already gone.